dynosaur_derive/src/lib.rs

```diff
@@ -134,41 +144,82 @@ fn blanket_impl_item(
             quote! {
-                type #ident #impl_generics = <Self as #trait_ident #trait_generics>::#ident #ty_generics #where_clause;
+                type #ident #impl_generics = #type_ref #ident #ty_generics #where_clause;
             }
         }
         _ => Error::new_spanned(item, "unsupported item type").into_compile_error(),
     }
 }
 
 fn mk_dyn_struct(struct_ident: &Ident, erased_trait: &ItemTrait) -> TokenStream {
-    let mut params: Punctuated<_, Token![,]> = Punctuated::new();
-    let mut param_args: Punctuated<_, Token![,]> = Punctuated::new();
-
     let erased_trait_ident = &erased_trait.ident;
-    params.push(quote! { 'dynosaur });
+    let (struct_params, trait_params) = struct_trait_params(erased_trait);
+
+    quote! {
+        struct #struct_ident #struct_params {
+            ptr: *mut (dyn #erased_trait_ident #trait_params + 'dynosaur_struct),
+            owned: bool,
+        }
+    }
+}
+
+fn struct_trait_params(erased_trait: &ItemTrait) -> (TokenStream, TokenStream) {
+    let mut struct_params: Punctuated<_, Token![,]> = Punctuated::new();
+    let mut trait_params: Punctuated<_, Token![,]> = Punctuated::new();
+
+    struct_params.push(quote! { 'dynosaur_struct });
     erased_trait.generics.params.iter().for_each(|item| {
-        params.push(quote! { #item });
-        param_args.push(quote! { #item });
+        struct_params.push(quote! { #item });
+        trait_params.push(quote! { #item });
     });
     erased_trait.items.iter().for_each(|item| match item {
         TraitItem::Type(TraitItemType { ident, .. }) => {
-            params.push(quote! { #ident });
-            param_args.push(quote! { #ident = #ident });
+            struct_params.push(quote! { #ident });
+            trait_params.push(quote! { #ident = #ident });
         }
         _ => {}
     });
 
-    let param_args = if param_args.is_empty() {
+    let trait_params = if trait_params.is_empty() {
         quote! {}
     } else {
-        quote! { <#param_args> }
+        quote! { <#trait_params> }
     };
 
+    (quote! { <#struct_params> }, trait_params)
+}
+
+fn mk_dyn_struct_impl_item(struct_ident: &Ident, erased_trait: &ItemTrait) -> TokenStream {
+    let erased_trait_ident = &erased_trait.ident;
+    let (_, trait_generics, where_clause) = &erased_trait.generics.split_for_impl();
+
+    let type_ref = quote! {};
+    let const_ref = quote! { <Self as #erased_trait_ident #trait_generics>:: };
+    let items = erased_trait.items.iter().map(|item| {
+        impl_item(item, &type_ref, &const_ref, |ident, args| {
+            let args = match &args[..] {
+                [arg, rest @ ..] => {
+                    if "self" == arg.to_string() {
+                        rest
+                    } else {
+                        &args
+                    }
+                }
+                _ => &args,
+            };
+
+            quote! {
+                unsafe { &*self.ptr }.#ident(#(#args),*).await
```

Comment:
I guess we will have to wait to make this work for all RPITIT. Seems like a bug upstream.